### ISSUE TITLE
PRD: Backup, Restore & Disaster Recovery (BCDR) Plan

### ISSUE THREAD

**Issue by jmservera** [jmservera/aithena#594]
## 🎯 Goal

Implement a comprehensive backup and restore system for all Aithena stateful data, with documented MTTR targets and a disaster recovery runbook. Exclude the book library itself (needs a separate large-file strategy); focus on everything else.

## 📋 Problem Statement

Aithena has **7 stateful data stores** and no backup system. A disk failure, misconfigured update, or accidental deletion could cause:
- **Auth DB loss**: All users locked out, credentials irrecoverable
- **Solr index loss**: Hours-to-days rebuild from source documents
- **Redis/RabbitMQ loss**: Indexing pipeline restarts from scratch
- **Secrets loss**: Cluster won't start without JWT secret, API keys
- **Collections DB loss** (v1.10+): Users lose curated reading lists and notes

## 📐 Stateful Data Inventory

| Data Store | Location (host) | Criticality | RPO Target | Can Rebuild? | Typical Size |
|------------|-----------------|-------------|------------|--------------|-------------|
| **Auth SQLite DB** | `$AUTH_DB_DIR` (~/.local/share/aithena/auth/) | 🔴 Critical | < 1 hour | ❌ No — credentials irreplaceable | < 1 MB |
| **Collections SQLite DB** (v1.10) | TBD `/data/collections/` | 🔴 Critical | < 1 hour | ❌ No — user-generated content | < 10 MB |
| **Runtime secrets** | `.env` file (not in git) | 🔴 Critical | RPO: 0 | ❌ No — JWT secret, API keys | < 1 KB |
| **Solr index** (3 nodes) | `/source/volumes/solr-data{1,2,3}` | 🟡 High | < 24 hours | ✅ Yes — rebuild from documents | 500 MB – 50+ GB |
| **ZooKeeper data** (3 nodes) | `/source/volumes/zoo-data{1,2,3}` | 🟡 High | < 24 hours | ✅ Yes — but quorum loss = Solr offline | < 100 MB |
| **RabbitMQ state** | `/source/volumes/rabbitmq-data` | 🟢 Medium | < 4 hours | ✅ Yes — document-lister re-scans | < 100 MB |
| **Redis state** | `/source/volumes/redis` | 🟢 Medium | < 4 hours | ✅ Yes — ephemeral processing state | < 100 MB |
| **SSL certificates** | `/source/volumes/certbot-data/` | 🟡 High | < 24 hours | ✅ Yes — certbot re-issues | < 1 MB |
| ~~Book library~~ | ~~`$BOOKS_PATH`~~ | ~~Excluded~~ | — | — | — |

### Failure Scenarios & Recovery Times

| Scenario | Impact | MTTR Estimate | Recovery Path |
|----------|--------|---------------|---------------|
| **Single service crash** (OOM, bug) | 1 service down | **< 1 min** | Docker auto-restart (`restart: unless-stopped`) |
| **Auth DB corrupted** | All users locked out | **5-15 min** | Restore from latest backup, restart solr-search |
| **Solr single-node failure** | Degraded search (2/3 replicas) | **< 5 min** | Docker restart, Solr auto-recovers from replicas |
| **Solr full cluster loss** | Search completely down | **4-24 hours** | Restore Solr + ZK volumes OR full reindex from documents |
| **Redis/RabbitMQ loss** | Indexing pipeline stalls | **5-30 min** | Restart services; document-lister re-scans library |
| **Full VM loss** | Everything down | **1-4 hours** | Provision new VM, restore all backups, restart cluster |
| **Secrets lost** | Cluster won't start | **15-30 min** | Restore .env from secure vault, restart |
| **Full disaster** (no backups) | Complete data loss | **Days** | Reprovision, recreate users manually, full reindex |

### Target SLA

| Metric | Target |
|--------|--------|
| **RTO** (Return to Operation) | < 1 hour for critical services, < 4 hours for full cluster |
| **RPO** (Recovery Point Objective) | < 1 hour for auth/collections, < 24 hours for Solr index |
| **Backup window** | < 5 min (must not block normal operations) |
| **Restore test frequency** | Monthly automated restore drill |

### 1. Backup System

**Owner: Brett (Infra)**

#### Backup Tiers

| Tier | Data | Frequency | Retention | Method |
|------|------|-----------|-----------|--------|
| **Critical** | Auth DB, Collections DB, .env secrets | Every 30 min | 7 days rolling | SQLite `.backup` API + encrypted copy |
| **High** | Solr + ZK volumes | Daily (off-peak) | 30 days rolling | Volume snapshot or tar + compress |
| **Medium** | Redis RDB, RabbitMQ | Daily | 14 days rolling | Copy RDB dump / volume snapshot |

#### Backup Script (`scripts/backup.sh`)

```bash
#!/bin/bash
# Aithena BCDR Backup
# Usage: ./scripts/backup.sh [--tier critical|high|all] [--dest /path/to/backups]

# Critical tier: SQLite online backup (no downtime)
sqlite3 /data/auth/users.db ".backup '/backups/auth-$(date +%Y%m%d-%H%M).db'"

# Encrypt sensitive backups
gpg --symmetric --cipher-algo AES256 --batch --passphrase-file /etc/aithena/backup.key ...

# High tier: Solr snapshot via API (no downtime)
curl "http://solr:8983/solr/admin/collections?action=BACKUP&name=books-$(date +%Y%m%d)&location=/backups/solr"

# Compress and ship to backup destination
```

#### Integration Options

**Option A: Cron-based (simplest)**
- `crontab` entry runs backup.sh on schedule
- Backups stored locally + rsync to remote

**Option B: Docker sidecar**
- New `backup` service in docker-compose
- Runs on schedule, accesses volumes via shared mounts
- Health check reports last successful backup

**Option C: Admin UI integration**
- Backup/restore controls on React admin page
- "Backup now" button, backup history, restore from backup
- Status indicator: "Last backup: 23 min ago ✅"

**Recommendation:** Start with **Option B** (Docker sidecar) + add **Option C** UI controls later. The sidecar is self-contained and doesn't require user intervention.

### 2. Restore System

**Owner: Brett (Infra) + Parker (Backend)**

#### Restore Script (`scripts/restore.sh`)

```bash
#!/bin/bash
# Aithena BCDR Restore
# Usage: ./scripts/restore.sh [--from /path/to/backup] [--component auth|solr|all]

# Pre-flight checks
# - Verify backup integrity (checksum)
# - Stop affected services
# - Create safety backup of current state before overwriting

# Restore auth DB
sqlite3 /data/auth/users.db ".restore '/backups/auth-latest.db'"

# Restore Solr (via collection API or volume replacement)
curl "http://solr:8983/solr/admin/collections?action=RESTORE&name=books&location=/backups/solr/books-latest"

# Post-restore verification
# - Auth: test login with known user
# - Solr: verify document count matches pre-backup
# - Redis/RabbitMQ: restart and verify connectivity
```

### 3. Admin UI Controls (v1.10.1+)

**Owner: Dallas (Frontend) + Parker (Backend)**

New section on the Admin page:

- **Backup status dashboard**: Last backup time per tier, success/failure, size
- **"Backup now" button**: Triggers immediate backup of selected tier
- **Backup history**: List of available backups with timestamps and sizes
- **Restore wizard**: Select backup → preview what will be restored → confirm → restore
- **Backup settings**: Configure schedule, retention, destination

#### API Endpoints

```
GET    /v1/admin/backups              # List available backups
POST   /v1/admin/backups              # Trigger immediate backup
GET    /v1/admin/backups/{id}         # Backup details (size, components, checksum)
POST   /v1/admin/backups/{id}/restore # Restore from specific backup
GET    /v1/admin/backups/status       # Last backup status per tier
PUT    /v1/admin/backups/config       # Update backup schedule/retention
```

### 4. Disaster Recovery Runbook

**Owner: Brett (Infra) + Newt (PM for docs)**

Document in admin manual:

1. **Assessment**: What failed? Single service, disk, or full VM?
2. **Triage decision tree**: Restart → restore → rebuild → reprovision
3. **Step-by-step restore procedures** per scenario
4. **Verification checklist** after restore
5. **Emergency contacts and escalation**
6. **Monthly drill procedure**

## 🏗️ Architecture

```
scripts/
  backup.sh           # Backup orchestrator
  restore.sh          # Restore orchestrator
  verify-backup.sh    # Integrity verification

src/solr-search/
  backup_service.py   # Backup API logic
  main.py             # New /v1/admin/backups/* endpoints

docker-compose.yml    # New backup sidecar service
  backup:
    image: aithena-backup
    volumes:
      - auth-data:/data/auth:ro
      - solr-data:/data/solr:ro
      - backup-dest:/backups
    environment:
      - BACKUP_SCHEDULE=*/30 * * * *
      - BACKUP_RETENTION_DAYS=30
      - BACKUP_ENCRYPTION_KEY_FILE=/run/secrets/backup-key
```

## 👥 Team Assignments

| Member | Role | Focus |
|--------|------|-------|
| **Brett** | Infra Architect (Lead) | Backup/restore scripts, Docker sidecar, volume management, DR runbook |
| **Parker** | Backend Dev | Backup API endpoints, SQLite backup integration, Solr snapshot API |
| **Dallas** | Frontend Dev | Admin UI backup controls (status, history, restore wizard) |
| **Kane** | Security | Backup encryption, secret handling, restore authorization |
| **Newt** | PM | DR runbook documentation, MTTR validation, drill scheduling |
| **Lambert** | Tester | Backup integrity tests, restore drill automation, chaos testing |

### Suggested Work Order
1. **Brett**: Backup/restore shell scripts (standalone, no Docker dependency)
2. **Brett**: Docker sidecar service for automated backups
3. **Parker**: API endpoints for backup status and trigger
4. **Kane**: Review encryption and access control
5. **Dallas**: Admin UI backup dashboard
6. **Lambert**: Automated restore drill test suite
7. **Newt**: DR runbook for admin manual

## ❓ Open Questions for Team Discussion

1. **Backup destination**: Local volume only, or also support remote (S3-compatible, NFS, rsync to remote host)? Local is simpler but doesn't protect against full disk loss.
2. **Solr backup method**: Solr Collections API BACKUP (clean, API-driven) vs volume-level tar (faster but requires quiescence)? Collections API is recommended.
3. **Encryption**: GPG symmetric (simple) vs age (modern, simpler key management)? Or defer encryption to the backup destination (encrypted filesystem)?
4. **Backup sidecar vs main service**: Should backup logic live in solr-search (already has access to everything) or a dedicated lightweight container?
5. **Admin UI scope for v1.10.1**: Full restore wizard, or just backup status + manual restore via CLI?
6. **Alerting**: Should the backup sidecar emit alerts (webhook, email) on backup failure? Or just expose status via the API?
7. **Collections DB path**: Where will the v1.10 collections SQLite DB live? Should it share `AUTH_DB_DIR` or have its own?
8. **Book library backup**: Out of scope for this issue, but should we create a separate issue for large-file backup strategy (dedup, incremental, offsite)?

## 🔗 Related
- Depends on: v1.10.0 collections feature (#591) — need to know DB location
- Related: Stress testing (#590) — backup window must not degrade performance
- Future: Book library backup strategy (separate issue, large-file problem)
- Future: Multi-site replication for HA deployments
